packages/react/src/useFetch.ts

```diff
@@ -0,0 +1,146 @@
+ * The second return value is a function for executing the fetch request. It returns a promise for the response body.
+ *
+ * By default, `GET` requests are sent as soon as the hook executes. Any other request methods are not sent automatically, and must be triggered by calling the `execute` function returned in the second argument.
+ *
+ * Pass the `{ json: true }` option to expect a JSON response from the server, and to automatically parse the response as JSON. Otherwise, the response will be returned as a `string` object.
+ *
+ * Pass the `{ stream: true }` to get a `ReadableStream` object as a response from the server, allowing you to work with the response as it arrives. Otherwise, the response will be returned as a `string` object.
+ *
+ * If you want to read model data, see the `useFindMany` function and similar. If you want to invoke a backend Action, use the `useAction` hook instead.
+ *
+ * @param path the backend path to fetch
+ * @param options the `fetch` options for the request
+ *
+ * @example
+ * ```
+ * export function UserByEmail(props: { email: string }) {
+ *   const [{data, fetching, error}, refresh] = useFetch("/users/get", {
+ *     method: "GET",
+ *     body: JSON.stringify({ email: props.email }})
+ *     headers: {
+ *       "content-type": "application/json",
+ *     }
+ *     json: true,
+ *   });
+ *
+ *   if (result.error) return <>Error: {result.error.toString()}</>;
+ *   if (result.fetching && !result.data) return <>Fetching...</>;
+ *   if (!result.data) return <>No user found with id={props.id}</>;
+ *
+ *   return <div>{result.data.name}</div>;
+ * }
+ */
+export function useFetch(path: string, options: { stream: true } & FetchHookOptions): FetchHookResult<ReadableStream<Uint8Array>>;
+export function useFetch<T extends Record<string, any>>(url: string, options: { json: true } & FetchHookOptions): FetchHookResult<T>;
+export function useFetch(path: string, options?: FetchHookOptions): FetchHookResult<string>;
+export function useFetch<T = string>(path: string, options?: FetchHookOptions): FetchHookResult<T> {
+  // Used to prevent state update if the component is unmounted
+  const mounted = useRef<boolean>(true);
+  const memoizedOptions = useStructuralMemo<FetchHookOptions>(options ?? {});
+  const connection = useConnection();
+
+  const [state, dispatch] = useReducer<Reducer<FetchHookState<T>, FetchAction<T>>, FetchHookOptions>(
+    reducer,
+    memoizedOptions,
+    (memoizedOptions) => {
+      return { fetching: startRequestByDefault(memoizedOptions), options: memoizedOptions };
+    }
+  );
+
+  const send = useCallback(
+    async (sendOptions?: Partial<FetchHookOptions>): Promise<T> => {
+      dispatch({ type: "fetching" });
+
+      let data: any;
+      let response: Response | undefined = undefined;
+
+      try {
+        // make the fetch call using GadgetConnection to pass along auth and other headers
+        response = await connection.fetch(path, { ...memoizedOptions, ...sendOptions });
```

Comment:
If `memoizedOptions?.json` is `true`, should we send `Accept: application/json` to give better guarantees about having a JSON response?

Reply:
Good call, added!